src/metatracer.rs

```diff
@@ -54,68 +94,131 @@ const PHASE_COUNTING: u32 = 0b10 << 30; // The value specifies the current hot c
+                                    pack.store(PHASE_COUNTING | count, Ordering::Relaxed);
+                                    break;
+                                },
+                                Err(e) => unreachable!("{}", e),
+                            }
+                        }
                     } else {
-                        new_pack = PHASE_COUNTING | (count + 1);
-                    }
-                    if pack.compare_and_swap(lp, new_pack, Ordering::Release) == lp {
-                        break;
+                        let new_pack = PHASE_COUNTING | (count + 1);
+                        if pack.compare_and_swap(lp, new_pack, Ordering::Release) == lp {
+                            break;
+                        }
                     }
                 },
                 PHASE_TRACING => {
-                    if pack.compare_and_swap(lp, PHASE_COMPILED, Ordering::Release) == lp {
+                    if trace.is_none() {
+                        // It wouldn't make sense to stop tracing multiple times, so stash the
+                        // trace first time around the pack update loop only.
+                        let mut res = Err(HWTracerError::Unknown);
+                        with_thread_tracer(&self.tracer, |tracer| res = tracer.stop_tracing());
+                        match res {
+                            Ok(t) => {
+                                // We've arrived back at the Location from which we started tracing
+                                // and the current thread is the one that started the tracing.
+                                trace = Some(t);
+                            },
+                            Err(HWTracerError::TracerState(TracerState::Stopped)) => {
+                                // Another thread is tracing this location. Do nothing.
+                                break;
+                            },
+                            Err(e) => panic!("Unexpected tracer error: {}", e),
+                        }
+                    }
+                    if pack.compare_and_swap(lp, PHASE_COMPILING, Ordering::Release) == lp {
+                        // Compilation is done in a background thread.
+                        let thr_pack = Arc::clone(&pack);
+                        let thr_code_cache = Arc::clone(&self.code_cache);
+                        thread::spawn(move || {
+                            let code = Code::compile(&trace.unwrap());
+                            let mut thr_code_cache_w = thr_code_cache.write().unwrap();
+                            // Store the index of where the new code exists in the code cache into
+                            // the value part of the pack so we can look it up in constant time.
+                            let idx = thr_code_cache_w.len() as u32;
+                            thr_code_cache_w.push(code);
+                            // We don't need a retry loop here. Once a location enters
```

Comment:
I don't understand this comment.

Reply:
My intuition is that no other thread can mutate the pack for the location once we start compiling there. So there's no possibility of a thread clash. But I'm willing to be proven wrong.